Add `keytype`+`valtype` method for AbstractArray and AbstractVector

tkluck opened the pull request:
The function `keytype(a)` returns `eltype(keys(a))` for associative structures; it should do the same for `AbstractArray` and `AbstractVector`. Looks like this was overlooked(?) in #25577.

#25104 is a related pull request, but renaming is less urgent than adding expected behaviour.

tkluck committed:
Add keytype method for AbstractArray and AbstractVector

The function `keytype(a)` returns `eltype(keys(a))` for associative
structures; it should do the same for `Abstract{Array, Vector}`.

Keno commented:
> eltype(keys(a))

Wouldn't literally this definition be a better generic fallback and cover these cases?

tkluck committed:
Add keytype method for AbstractArray and AbstractVector (part 2)

Now also for the type, not just for objects.

andyferris commented:
I feel the fallback can be allowed to have allocations, whether or not the compler can elide them. If we have specific performance problems with `Base` types we can simply add some specializations.

> And we'd need to implement it case-by-case for `keytype` on the type anyway, as we can't call `keys` on the type.

This is more convincing.

tkluck committed:
Merge remote-tracking branch 'origin/master' into keytype-abstractarray

Merge back master branch to validate clean merge and to kick off
an up-to-date CI build.

tkluck committed:
Add valtype(...) for arrays

Arrays are associative structures in the sense that keys(...)
and values(...) both work on them. When considering them as such, their
valtype should be equal to eltype. (A future version of Julia could
even do `const valtype = eltype`?)

tkluck commented:
Reviving this pull request as I recently found myself reaching for this. Merged back master (still a clean merge) and added `valtype`.

tkluck commented on base/abstractarray.jl:
```diff
@@ -96,6 +96,9 @@ unsafe_indices(r::AbstractRange) = (OneTo(unsafe_length(r)),) # Ranges use check
 keys(a::AbstractVector) = LinearIndices(a)
 
+keytype(a::AbstractArray) = CartesianIndex{ndims(a)}
+keytype(a::AbstractVector) = Int
```
Not 100% sure, but a look at `SparseVector` seems to indicate that it's supposed to be `Int` even if I try my best to make it `Int16`:
```
julia> dump(SparseVector(Int16(3), Int16[2,3], [1.0,2.0]))
SparseVector{Float64,Int16}
  n: Int64 3
  nzind: Array{Int16}((2,)) Int16[2, 3]
  nzval: Array{Float64}((2,)) [1.0, 2.0]

julia> dump(eachindex(SparseVector(Int16(3), Int16[2,3], [1.0,2.0])))
Base.OneTo{Int64}
  stop: Int64 3
```
And also [the custom indices docs](https://docs.julialang.org/en/v1/devdocs/offset-arrays/) make no mention of allowing arbitrary index types. My mental model is "`Int` corresponds to `size_t`".

mbauman replied:
There are some types with non-`Int` axes but it's not a very well-trod ground, so I'd anticipate some potholes.

```julia
julia> r = typemax(Int128)-10:typemax(Int128)
170141183460469231731687303715884105717:170141183460469231731687303715884105727

julia> dump(length(r))
Int128 11

julia> dump(eachindex(r))
Base.OneTo{Int128}
  stop: Int128 11
```

tkluck replied:
Aha, thanks for finding that example @mbauman . Let me also explicitly add that
```
julia> r isa AbstractArray
true
```
(which you implied already and) which is why this is relevant.

OTOH, [the array interface](https://docs.julialang.org/en/v1/manual/interfaces/#man-interface-array-1), actually does specify `Int` as the indexing type. More precisely, it makes a distinction between `IndexLinear` and `IndexCartesian` (and I should definitely add that distinction to this pull request) and prescribes a `getindex(A, i::Int)` method in the former case.

I think the sanest way forward is:

1. extend the [indexing interface](https://docs.julialang.org/en/v1/manual/interfaces/#Indexing-1) with required `keytype` and `valtype` methods
2. extend the array interface with `::Integer` and `::VarArg{<:Integer,N}` methods for `IndexLinear` and `IndexCartesian` respectively. (In the latter case, different axes may have different integer types. An interesting one is `Bool <: Integer`.)
2. do a sweep of `Base` to add these.

I'm not looking forward particularly to that last bit, but I can probably squeeze in the time if we think this is what should be merged.

mbauman replied:
I don't see this as nearly as big of a roadblock here.

1. I think `Int` and `CartesianIndex` are perfectly sane defaults that can be automatically provided for everyone.
2. You can _already_ just define your indexing methods in terms of `::Integer` without any trouble.  But if you let it, the Base fallbacks will happily convert them to `Int`s for you and call your `::Int` method.  We really don't have a demand for folks wanting to define `::Int8` indexing and finding it annoying to define both `getindex(A, I::Integer) = A[Int8(I)]` and their canonical `::Int8` method.

There's a distinction between the _implementation_ of these arrays (either `IndexLinear` or `IndexCartesian`) and their `keys`.  The former is `eachindex` (and can really be anything that's fast), whereas the latter is `keys` and defined to be either the `LinearIndices` or the `CartesianIndices` depending on the dimensionality.  And note that those two structures are (currently) completely limited to being `Int`-based.

tkluck replied:
@mbauman thanks for adding that clarity. Under this understanding, I guess the pull request is good-to-go as-is, no? Let's wait for @andyferris' opinion before resolving.

andyferris replied:
If @mbauman is OK than it's definitely OK by me!

mbauman commented on base/abstractarray.jl:
```diff
@@ -101,6 +101,15 @@ unsafe_indices(r::AbstractRange) = (OneTo(unsafe_length(r)),) # Ranges use check
 keys(a::AbstractArray) = CartesianIndices(axes(a))
 keys(a::AbstractVector) = LinearIndices(a)
 
+keytype(a::AbstractArray) = CartesianIndex{ndims(a)}
```
```suggestion
keytype(a::AbstractArray) = keytype(typeof(a))
```

mbauman commented on base/abstractarray.jl:
```diff
@@ -101,6 +101,15 @@ unsafe_indices(r::AbstractRange) = (OneTo(unsafe_length(r)),) # Ranges use check
 keys(a::AbstractVector) = LinearIndices(a)
 
+keytype(a::AbstractArray) = CartesianIndex{ndims(a)}
+keytype(a::AbstractVector) = Int
```
```suggestion
```

mbauman commented on base/abstractarray.jl:
```diff
@@ -101,6 +101,15 @@ unsafe_indices(r::AbstractRange) = (OneTo(unsafe_length(r)),) # Ranges use check
+keytype(A::Type{<:AbstractArray}) = CartesianIndex{ndims(A)}
+keytype(A::Type{<:AbstractVector}) = Int
+
+valtype(a::AbstractArray) = eltype(a)
```
```suggestion
valtype(a::AbstractArray) = valtype(typeof(a))
```

mbauman reviewed (commented):
Let's make the type-based methods the "canonical" endpoints so a custom type would only need to override that one (instead of both).

mbauman committed:
Implement keytype/valtype for object as keytype(typeof(...

As suggested by @mbauman in code review.

Co-Authored-By: tkluck <tkluck@infty.nl>

tkluck committed:
Implement keytype/valtype for ranges

tkluck committed:
Add test cases for keytype/valtype(::AbstractArray)

tkluck committed:
Add #27749 to NEWS.md

tkluck commented:
All feedback is now incorporated, including the needs-* labels. The CI failures seem to be about dependency artifacts and not related to this PR.

tkluck commented:
Oooh let's get this in 1.2: https://discourse.julialang.org/t/julia-1-2-feature-freeze-april-4th/22478

@mbauman or @andyferris , do either of you have commit rights? Or should we bring in someone else?

StefanKarpinski commented:
It appears that this no longer needs NEWS or tests but still needs docs with compatibility annotation indicating that it was added in version 1.2. If @mbauman approves, we can merge. It would be great to get the docs with this PR but if time is an issue if you promise to add docs after the feature freeze...

tkluck committed:
keytype/valtype for arrays: docs + compat annotation

mbauman reviewed (approved):
Thanks @tkluck! This looks great to me now. Let's merge once tests pass.

tkluck commented:
@StefanKarpinski good catch, apologies. (FWIW, I (mis)understood that that label was related to the Compat.jl package, and that it would need the specific commit that introduced this feature -- i.e. that it wasn't actionable until this was merged.)

Just pushed those docs with a compatibility note.

StefanKarpinski commented on base/abstractarray.jl:
````diff
@@ -101,6 +101,51 @@ unsafe_indices(r::AbstractRange) = (OneTo(unsafe_length(r)),) # Ranges use check
+
+# Examples
+```jldoctest
+julia> valtype(Int32[1, 2, 3])
````
```suggestion
julia> valtype(["one", "two", "three"])
```

StefanKarpinski commented on base/abstractarray.jl:
````diff
@@ -101,6 +101,51 @@ unsafe_indices(r::AbstractRange) = (OneTo(unsafe_length(r)),) # Ranges use check
+# Examples
+```jldoctest
+julia> valtype(Int32[1, 2, 3])
+Int32
````
```suggestion
String
```

StefanKarpinski commented:
Looks great! I made a suggestion to change the `valtype` example to use a string array since it seemed more obviously different than the types of the keys of an array which could be `Int32` on some systems.

StefanKarpinski committed:
keytype/valtype: use String values in examples

this makes the distinction between key and value clearer. Thanks to @StefanKarpinski for suggesting this.

Co-Authored-By: tkluck <tkluck@infty.nl>

tkluck commented:
Travis uncovers failing

    keytype_is_correct(Int64(1):Int64(5))

on 32-bits architecture. I'll check.

tkluck committed:
keytype/valtype for arrays: no special treatment for ranges

This was a little red herring in the discussion in #27749:
the `length` of a range may depend on the type of its parameters,
but as it turns out, `eltype(keys(...))` does not.

That's arguably a bug/inconsistency, but it's outside of the
scope of this pull request to fix that.

tkluck committed:
Merge remote-tracking branch 'origin/master' into keytype-abstractarray

tkluck commented:
buildbot is happy now, including the new test cases introduced by this pull request. Appveyor and travis complain about unrelated things and are broken on master too.

I say, good to go :)

tkluck committed:
Merge branch 'master' into keytype-abstractarray

StefanKarpinski reviewed (approved):
Good to go as soon as CI passes. Please squash when merging!
